### ISSUE TITLE
[PERFORMANCE] OmniMoE_Torch forward function issue

### ISSUE THREAD

**Issue by tusiqi1** [flash-algo/omni-moe#19]
### Performance issue description

replace 
        scores_x, indices_x = router_log_probs_x.topk(self.num_expert_sqrt, dim=-1)
        scores_y, indices_y = router_log_probs_y.topk(self.num_expert_sqrt, dim=-1)
as 
        scores_x, indices_x = router_log_probs_x.topk(min(self.num_expert_sqrt, self.topk), dim=-1)
        scores_y, indices_y = router_log_probs_y.topk(min(self.num_expert_sqrt, self.topk), dim=-1)

### Current performance metrics

None

### Environment information

```shell
None
```

### Benchmark code

```python

```

**Comment by LoserCheems** [comment]
Great trick! Taking `topk(min(num_expert_sqrt, top_k))` per axis instead of the full `num_expert_sqrt` is a clean and correct optimization — the proof is straightforward: any expert in the global top-K must have both its x- and y-indices within the top-K of their respective axes, so reducing the candidate pool this way is lossless.

We've verified it numerically (max absolute error = 0 in float32). Would love to see this land in a PR — happy to assist if needed!

**Comment by tusiqi1** [comment]
I have another question: for a decoder-only Transformer architecture, during multi-batch SFT training, it seems that the PyTorch implementation does not account for the impact of padding when using BatchNorm.
